design/ceph/object/object-store-backend.md

```diff
@@ -0,0 +1,103 @@
+---
+title: Object Store Backend
+target-version: release-1.10
+---
+
+# Overview
+
+Ceph RADOS Gateway is undergoing a major overhaul with the Zipper project. The goal of Project
+Zipper is to create an API that can be used to separate the RGW top end (represented by Swift and
+S3) from bottom end providers (represented currently by RADOS, and eventually by Cloud, File, and
+Memory backends). This should allow the front ends to access any implemented back end
+transparently.
+
+The secondary objective is to allow multiple layers of these API providers to stack up, allowing
+transformations of operations (such as redirection to different backends), or caching layers that
+cut across multiple backends.
+
+The primary Store is the current RadosStore, which provides access to a Ceph cluster via RADOS.
+However, new Stores are possible that store the data in any desired platform. One such Store, called
+DBStore, has been developed that stores data in SQL, and specifically in a local SQLite database.
+
+## Goals
+
+Rook-Ceph has a tremendous support for the Ceph Object Store and we want to leverage that.
+Especially the various controllers that are reconciling ObjectStore, Multisite configurations and
+Bucket provisioning.
+Mainly we are looking at providing the next generation of RGW via Rook.
+
+# Proposal details
+
+The `CephObjectStore` API will have a few details added to its spec such as an image so that it can
+be deployed without a Ceph cluster.
+
+```go
+type ObjectStoreSpec struct {
+...
+	// Image is the name of the image to use to deploy the object store
+	Image string `json:"image"`
+...
+}
+```
+
+Also, we need to expose a new type for the backend to use:
+
+```go
+type GatewaySpec struct {
+...
+	// Backend refers to the backend to use for the object store
+	// +optional
+	Backend *ObjectStoreBackend `json:"backend,omitempty"`
+...
+}
+```
+
+Where the `ObjectStoreBackend` is represented as such:
+
+```go
+type ObjectStoreBackend struct {
+	// The name of the backend
+	// +optional
+	// +kubebuilder:validation:Enum={"rados","sqlite"}
+	Name string `json:"name,omitempty"`
+
+    // Config is the configuration for the backend
+	// +optional
+    Config map[string]string `json:"config,omitempty"`
+}
+```
+
+Then the `CephObjectStore` `spec.go` file will need to be adapted depending on the backend.
+
+## Risks and Mitigation
+
```

Comment:
I am not sure how well the adminops will interact with other backend vendors

Reply:
I think it's up to the vendor to comply to the admin ops API but it's a valid point of concern, thanks for raising it.